mysql-test/suite/galera/t/galera_backup_stage.test

```diff
@@ -0,0 +1,83 @@
+#
+# Check that BACKUP STAGE BLOCK_DDL desyncs and pauses the node until BACKUP STAGE END:
+# - Local DDLs will fail immediately
+# - Local DMLs will block until resync
+# - Remote txns will be applied after resync (STAGE END).
+# 
+
+--source include/galera_cluster.inc
+--source include/have_innodb.inc
+
+--connection node_1
+CREATE TABLE t1 (f1 varchar(10)) ENGINE=InnoDB;
+
+# First, check that BACKUP STAGE END skipping desyncing stages is fine
+BACKUP STAGE START;
+BACKUP STAGE FLUSH;
+BACKUP STAGE END;
+
+BACKUP STAGE START;
+BACKUP STAGE FLUSH;
+
+--connect node_1a, 127.0.0.1, root, , test, $NODE_MYPORT_1
+--connection node_1a
+INSERT INTO t1 (f1) values ("node1_1");
+ALTER TABLE t1 ADD COLUMN (f2 int(10));
+
+--connection node_2
+INSERT INTO t1 (f1) values ("node2_1");
+ALTER TABLE t1 ADD COLUMN (f3 int(10));
+
+--connection node_1
+# BLOCK_DDL desyncs and pauses the node
+BACKUP STAGE BLOCK_DDL;
+
+--connection node_2
+INSERT INTO t1 (f1) values("node2_2");
+ALTER TABLE t1 ADD COLUMN (f5 int(10));
+
+--connection node_1a
+--error 1047
+ALTER TABLE t1 ADD COLUMN (f4 int(10));
+--send INSERT INTO t1 (f1) values("node1a");
+
```

Comment:
--send is asynchronous operation and can cause unpredictable effects in the test. There is no guarantee of how far the INSERT execution has reached or if it even has started to execute, before following mtr test command is executed (BACKUP STAGE BLOCK COMMIT, in this case).
Best approach is always to use a control connection, which monitors the progress of the "sent" statement until it has reached the wanted blocking point. In this case, if you can figure out the state where this INSERT will block, and add an wait_condition to enforce that INSERT has reached this, please add this check to the test.
Sometimes, it is not necessary to wait for --send operations, if it obvious from the test structure, that it does not matter in which exact execution stage the sent operation is at the time of next test script command. If this is the case, in this test, please add a comment to tell that INSERT is not waited for, it will help the next mtr test developer who is looking at this

Reply:
Could not find a way to ensure the insert indeed gets blocked inside sync_wait in the provider, but now it should be very unlikely it doesn't. In any case, even if it doesn't reach the provider sync_wait before STAGE END, test result is still the same (inserts complete after resync)